dartagnan/src/main/java/com/dat3m/dartagnan/solver/caat4wmm/RefinementModel.java

```diff
@@ -0,0 +1,126 @@
+package com.dat3m.dartagnan.solver.caat4wmm;
+
+
+import com.dat3m.dartagnan.wmm.Assumption;
+import com.dat3m.dartagnan.wmm.Constraint;
+import com.dat3m.dartagnan.wmm.Relation;
+import com.dat3m.dartagnan.wmm.Wmm;
+import com.dat3m.dartagnan.wmm.analysis.RelationAnalysis;
+import com.dat3m.dartagnan.wmm.axiom.ForceEncodeAxiom;
+import com.dat3m.dartagnan.wmm.utils.ConstraintCopier;
+import com.dat3m.dartagnan.wmm.utils.Cut;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+public class RefinementModel {
+
+    private final Wmm baseModel;
+    private final Wmm originalModel;
+    private final BiMap<Constraint, Constraint> orig2BaseConstraints;
+    private final BiMap<Relation, Relation> orig2BaseRelations;
+
+    public Wmm getBaseModel() { return baseModel; }
+
+    public Wmm getOriginalModel() { return originalModel; }
+
+    public Relation translateToBase(Relation rel) {
+        return orig2BaseRelations.get(rel);
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T extends Constraint> T translateToBase(T constraint) {
+        return (T)orig2BaseConstraints.get(constraint);
+    }
+
+    public Relation translateToOriginal(Relation rel) {
+        return orig2BaseRelations.inverse().get(rel);
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T extends Constraint> T translateToOriginal(T constraint) {
+        return (T)orig2BaseConstraints.inverse().get(constraint);
+    }
+
+    public Set<Relation> computeBoundaryRelations() {
+        final Set<Relation> boundary = new HashSet<>();
+        for (Constraint c : originalModel.getConstraints()) {
+            if (orig2BaseConstraints.containsKey(c)) {
+                continue;
+            }
+            c.getConstrainedRelations()
+                    .stream().map(orig2BaseRelations::get)
+                    .filter(Objects::nonNull)
+                    .forEach(boundary::add);
+        }
+
+        return boundary;
+    }
+
```

Comment:
Yes, it is related to the `Cut` boundary. From the refinement perspective, the boundary relations are those that get eagerly encoded (cut) but are also needed for the lazy part (cause it does compute derived relations from them).
But to make sure that they actually get encoded, we need some axiom on top of them (for otherwise the active sets will be empty), so we place an artificial `ForceEncodeAxiom`.

Reply:
Agree ... I have seen quite a few cases where RA on the whole model takes quite a lot of time, so better to avoid it when possible